File: src/controllers/_save_research_state.py
from controllers.__action import Action
from factorio_instance import PLAYER
from models.research_state import ResearchState
from models.technology_state import TechnologyState


class SaveResearchState(Action):
    def __init__(self, connection, game_state):
        super().__init__(connection, game_state)

    def __call__(self) -> ResearchState:
        """
        Save the current research state of the force

        Returns:
            ResearchState: Complete research state including all technologies
        """
        state, _ = self.execute(PLAYER)

        if not isinstance(state, dict):
            raise Exception(f"Could not save research state: {state}")

        try:
            # Convert the raw state into our dataclass structure
            technologies = {}
            if "technologies" in state:
                technologies = {
                    name: TechnologyState(
                        name=tech["name"],
                        researched=tech["researched"],
                        enabled=tech["enabled"],
                        level=tech["level"],
                        research_unit_count=tech["research_unit_count"],
                        research_unit_energy=tech["research_unit_energy"],
                        prerequisites=tech["prerequisites"],
                        ingredients=tech["ingredients"]
                    )
                    for name, tech in state["technologies"].items()
                }
            return ResearchState(
                technologies=technologies,
                current_research=state["current_research"] if "current_research" in state else None,
                research_progress=state["research_progress"],
                research_queue=state["research_queue"]
            )

        except Exception as e:
            print(f"Could not save technologies: {e}")
            raise e

File: src/actions/save_research_state.lua
global.actions.save_research_state = function(player_index)
    local player = game.get_player(player_index)
    local force = player.force

    -- Helper to serialize technology state
    local function serialize_technology(tech)
        local prerequisites = {}
        for name, _ in pairs(tech.prerequisites) do
            table.insert(prerequisites, "\""..name.."\"")
        end

        local ingredients = {}
        for _, ingredient in pairs(tech.research_unit_ingredients) do
            table.insert(ingredients, {
                name = "\""..ingredient.name.."\"",
                amount = ingredient.amount
            })
        end

        return {
            name = "\""..tech.name.."\"",
            researched = tech.researched,
            enabled = tech.enabled,
            -- visible = tech.visible,
            level = tech.level,
            research_unit_count = tech.research_unit_count,
            research_unit_energy = tech.research_unit_energy,
            prerequisites = prerequisites,
            ingredients = ingredients
        }
    end

    local research_state = {
        technologies = {},
        current_research = nil,
        research_progress = 0,
        research_queue = {}
    }

    -- Save all technology states
    for name, tech in pairs(force.technologies) do
        research_state.technologies[name] = serialize_technology(tech)
    end

    -- Save current research and progress
    if force.current_research then
        research_state.current_research = "\""..force.current_research.name.."\""
        research_state.research_progress = force.research_progress
    end

    -- Save research queue if it exists
    if force.research_queue then
        for _, tech in pairs(force.research_queue) do
            table.insert(research_state.research_queue, "\""..tech.name.."\"")
        end
    end

    return research_state
end